pkg/gentrait/securityevent/model_server.go

```diff
@@ -0,0 +1,73 @@
+package securityevent
+
+import (
+	"context"
+	"strconv"
+
+	"google.golang.org/grpc"
+
+	"github.com/smart-core-os/sc-golang/pkg/resource"
+	"github.com/vanti-dev/sc-bos/pkg/gen"
+)
+
+type ModelServer struct {
+	gen.UnimplementedSecurityEventApiServer
+	model *Model
+}
+
+func NewModelServer(model *Model) *ModelServer {
+	return &ModelServer{model: model}
+}
+
+func (m *ModelServer) Register(server *grpc.Server) {
+	gen.RegisterSecurityEventApiServer(server, m)
+}
+
+func (m *ModelServer) Unwrap() any {
+	return m.model
+}
+
+func (m *ModelServer) ListSecurityEvents(_ context.Context, req *gen.ListSecurityEventsRequest) (*gen.ListSecurityEventsResponse, error) {
+
+	// page token is just the index of where we left off (if any)
+	pageToken := req.GetPageToken()
+	startIndex := len(m.model.allSecurityEvents)
+	if pageToken != "" {
+		_, err := strconv.Atoi(req.GetPageToken())
+		if err != nil {
+			return nil, err
+		}
+		startIndex, _ = strconv.Atoi(pageToken)
+	}
+
+	count := req.PageSize
+	if count == 0 {
+		count = 50
+	} else if count > 1000 {
+		count = 1000
+	}
+
+	var resp gen.ListSecurityEventsResponse
+
+	// reverse to retrieve the latest events first
+	for i := startIndex - 1; i >= 0; i-- {
+		resp.SecurityEvents = append(resp.SecurityEvents, m.model.allSecurityEvents[i])
+		if len(resp.SecurityEvents) >= int(count) {
+			resp.NextPageToken = strconv.Itoa(i - 1)
```

Comment:
Normally I'd say using an index as a page token isn't safe, indexes tend to change and you can get duplicated or missing items between pages if the list grows/shrinks between requests. I _think_ it's safe in this case though because the index starts at the oldest event so shouldn't change. Might be worth a comment line somewhere to let future us know why it's safe here.